Fix wrong report on closure args mismatch when a ref is expected but not found

Emilgardis opened the pull request:
Fixes #42143.

Also adds test for this case.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

arielb1 commented:
@Emilgardis 

I would personally prefer to issue the error message when you find that the obligation & closure *trait refs* are tuples of different arg-counts, rather than trying to match against type errors.

nikomatsakis commented:
I'm not sure if this is the same thing that @arielb1 was saying, but I feel like we could do a lot better with this error message. That is, ideally, we would tell the user that the closure is declared with argument type `(u32, u32)`, but the iterator is providing a `&(u32, u32)`. It's a bit tricky to find just the right phrasing here but I think we can do it. :) @Emilgardis -- are you up for trying to go the extra mile here?

I think the error message we want is probably something like this:

```
error[E0593]: type mismatch in closure arguments
 --> src/main.rs:4:30
  |
4 |     let b: Vec<_> = a.iter().map(|x: (u32, u32)| 45).collect(); 
  |                              ^^^ ------------------ takes a `(u32, u32)`
  |                              |
  |                              expected closure that takes a `&(u32, u32)`
```

nikomatsakis commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -684,24 +684,36 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                                    expected_trait_ty.is_closure())
                 } else if let &TypeError::Sorts(ref expected_found) = e {
-                    let expected = if let ty::TyTuple(tys, _) = expected_found.expected.sty {
+                    let expected_len = if let ty::TyTuple(tys, _) = expected_found.expected.sty {
```
Huh. I am a bit dubious about this `if` in the first-place. I guess I have to run the code -- what happens if you just remove this logic altogether, though? It feels very odd to me to have to do all this matching.

nikomatsakis replied:
(In other words, the logic here feels quite suspect, even before your changes.)

Emilgardis commented:
I agree with what has been commented here. I'll get to the implementation most likely tomorrow. Should I open a new pr for it or just force push a new commit?

Mark-Simulacrum commented:
Force push is probably fine in this case since changes are small; in general just a normal push would be best

nikomatsakis commented:
@Emilgardis force push is good -- thanks!

carols10cents commented:
Hi @Emilgardis! Just a friendly ping to keep this on your radar :)

nikomatsakis commented:
@Emilgardis do you think you know how to do the better error message, or do you want a few tips?

Emilgardis commented:
Sorry for the delay, I have a unpushed commit on my local branch, I've removed the match @nikomatsakis suspected to be uneeded. So far no bugs regarding that change.

I've also implemented E0604, and I'm now trying to get a correct output of the types that where mismatched. i.e show `(u32, u32)` instead of `std::ops::FnOnce<((u32, u32),)>`

Emilgardis commented:
I realised that the error:

```
[E0604]: type mismatch in closure arguments
 --> src/main.rs:4:30
  |
4 |     let b: Vec<_> = a.iter().map(|x: (u32, u32)| 45).collect(); 
  |                              ^^^ ------------------ takes a `(u32, u32)`
  |                              |
  |                              expected closure that takes a `&(u32, u32)`
```
Is awfully closely related to E0281, should I remove my local E0604 and just change E0281 to give prettier output and have it be called on code like the example above instead of E0604?

arielb1 commented:
Do you need any help @Emilgardis? Feel free to ask me on IRC (I'm arielb1 or arielby).

nikomatsakis commented:
@Emilgardis it's always a tough call where to use a distinct error number. I think in this case it is worth keeping a distinct error number, though. Basically this is because we can give a more specialized --explain description targeting the particular case of closure arguments. Also, there *is* a sort of distinction here: the more general mismatched types refers to the value itself, whereas this is saying that the types in the trait that the closure implements don't match.

(Also, the current infrastructure makes it difficult to combine error messages in any case.)

nikomatsakis commented:
@Emilgardis sounds like you have something working? I guess you didn't yet push those commits? The preview looks very appealing though!

arielb1 commented:
@Emilgardis I couldn't find you on IRC. Are you still interested in working on this? Should I get this done myself?

Emilgardis commented:
Sorry for not being active, I'll get to this tomorrow(in 7-10 hours). I'm Emilgardis on mozilla

aidanhs commented:
Hi @Emilgardis, did you get a chance to come back to this?

Emilgardis commented:
Yes, I have a working branch on [temp-issue-42143](https://github.com/Emilgardis/rust/tree/temp-issue-42143). I'll get to rebasing it into changes upstream today

nikomatsakis commented:
@Emilgardis awesome :)

Emilgardis commented:
There we go. Should be all fine.

I'll address one concern I have in the code-comments.

Emilgardis commented on src/test/ui/mismatched_types/closure-arg-type-mismatch.stderr:
```diff
@@ -0,0 +1,36 @@
+  --> $DIR/closure-arg-type-mismatch.rs:16:27
+   |
+16 |         let d1 = a.iter().map(|x: &(u16,u16)| 45);
+   |                           ^^^ ------------------ takes a `for<'r> std::ops::FnMut<(&'r (u16, u16),)>`
```
Should this binder/lifetime be displayed?

Emilgardis commented on src/test/ui/mismatched_types/closure-arg-type-mismatch.stderr:
```diff
@@ -0,0 +1,36 @@
+15 |         let b = a.iter().map(|x: (u32, u32)| 45);
+   |                          ^^^ ------------------ takes a `std::ops::FnMut<((u32, u32),)>`
+   |                          |
+   |                          expected closure that takes a `std::ops::FnMut<(&(u32, u32),)>`
```
I'm thinking about implementing a hint telling the user to add a `&`/`&mut` to the closure

Emilgardis replied:
Also, I couldn't find a way (except for "parsing") to convert `Fn*<([&]({T} ... ))>` (meaning what is displayed here as `FnMut<(&(u32, u32),)>`) into a simple `&(u32, u32)` as @nikomatsakis suggested.

Emilgardis commented:
I'm not sure how to proceed. E0281 is very similar to E0623. Infact, what I can see the two errors are almost always the same.

nikomatsakis commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -825,6 +820,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+        let mut err = struct_span_err!(self.tcx.sess, span, E0623,
+            "type mismatch in closure arguments");
+        if let Some(sp) = found_span {
+            err.span_label(span, format!("expected closure that takes a `{}`", found));
```
we should be able to extract the types of the arguments by examining the **substs** of the `expected_ref` and `found` trait-refs. Something like:

```
let expected_args = expected_ref.map_bound_ref(|trait_ref| trait_ref.substs.type_at(0));
```

But it might need to be `type_at(1)`. That will give you a `Binder<Ty<'tcx>>`, which you should be able to use with `{}` I think.

nikomatsakis commented:
@Emilgardis ping?

arielb1 commented:
@Emilgardis ping. I'll be on vacation for the rest of the week, but if you need any help you can talk with @nikomatsakis .

arielb1 commented:
Closing due to inactivity. Feel free to reopen when you can find me.
